src/types/mfi.cpp

```diff
@@ -173,3 +173,130 @@ bool ReadMFI (MemFile &file, std::shared_ptr<Disk> &disk)
+			return heads == 2 ? DSSD : SSSD;
+		case Encoding::Amiga:
+		case Encoding::MFM:
+			switch (track[0].datarate)
+			{
+			case DataRate::_250K:
+			case DataRate::_300K:
+				return cyls > 40 ? DSQD : DSDD;
+			case DataRate::_500K:
+				return DSHD;
+			case DataRate::_1M:
+				return DSED;
+			default:
+				break;
+			}
+			break;
+		default:
+			break;
+		}
+	}
+
+	return SSSD;
+}
+
+
+bool WriteMFI (FILE* f_, std::shared_ptr<Disk> &disk)
+{
+#ifndef HAVE_ZLIB
+	throw util::exception("MFI disk images are not supported without ZLIB");
+#else
+	int tracks, heads;
+
+	std::vector<uint8_t> header(sizeof(MFI_FILE_HEADER), 0);
+	auto &fh = *reinterpret_cast<MFI_FILE_HEADER*>(header.data());
+	auto &track0 = disk->read_track({ 0, 0 });
+
+	tracks = static_cast<uint32_t>(disk->cyls());
+	heads = static_cast<uint32_t>(disk->heads());
+
+	strncpy(fh.signature, "MESSFLOPPYIMAGE", sizeof(fh.signature));
+	fh.cyl_count = util::htole(tracks);
+	fh.head_count = util::htole(heads);
+	fh.form_factor = util::htole(FF_UNKNOWN);
+	fh.variant = util::htole(MfiVariant(track0, disk->cyls(), disk->heads()));
+
+	if (!fwrite(header.data(), header.size(), 1, f_))
+		throw util::exception("write error");
+
+	std::map<CylHead, MFI_TRACK_HEADER> track_lut;
+	std::map<CylHead, FluxData> bitstreams;
+	int pos = sizeof(MFI_FILE_HEADER) + tracks * heads * sizeof(MFI_TRACK_HEADER);
+
+	if (fseek(f_, pos, SEEK_SET) < 0)
+		throw util::exception("seek error");
+
+//
+
+	auto max_disk_track_bytes = 0;
+	for (int cyl = 0; cyl < tracks; ++cyl)
+	{
+		auto max_track_bytes = 0;
+		for (int head = 0; head < heads; ++head)
+		{
+			CylHead cylhead(cyl, head);
+			auto trackdata = disk->read(cylhead);
+			auto bitstream = trackdata.preferred().flux();
+			int track_bytes = bitstream[0].size() * 4;
+			max_track_bytes = std::max(track_bytes, max_track_bytes);
+			max_disk_track_bytes = std::max(max_disk_track_bytes, max_track_bytes);
+			bitstreams[cylhead] = std::move(bitstream);
+		}
+	}
+
+	for (int cyl = 0; cyl < tracks; ++cyl)
+	{
+		for (int head = 0; head < heads; ++head)
+		{
+			CylHead cylhead(cyl, head);
+			auto &bitstream = bitstreams[cylhead];
+			auto track_size = bitstream[0].size() + 1;
+			int orient = 0, total = 0;
+
+			std::vector<uint32_t> track_data(track_size - 1);
+			Data compressed_data(track_size * 4 + 1000);
+
+			std::transform(bitstream[0].begin(), bitstream[0].end(), track_data.begin(),
+				[&orient, &total](uint32_t a) -> uint32_t { orient ^= 1; total += a; return (a) | (orient ? MG_B : MG_A); });
+			track_data.push_back((200000000 - total) | (orient ? MG_B : MG_A));
+
+			uLongf csize = track_size * 4 + 1000;
+			int rc = compress(&compressed_data[0], &csize, (const Bytef *)&track_data[0], track_size * 4);
+			if (rc != Z_OK) {
+				util::cout << "compress of " << CylHead(cyl, head) << " failed, rc " << rc << "\n";
+				return false;
+			}
+
+			track_lut[cylhead] = { util::htole(pos), util::htole(csize), util::htole(track_size * 4), 0 };
```

Comment:
This generates a narrowing warning on all compilers I've tried.  pos, csize, and track_size all need static_casting to uint32_t to match the MFI header types.  Please ensure the build is warning-free on the compiler you're using.

Reply:
I'll merge it in anyway as Visual Studio is complaining about a std::accumulate case that gcc seems happy with.  I'll look into that separately.